database/transaction.rs

```diff
@@ -80,24 +83,19 @@ impl<D: DurabilityClient> TransactionWrite<D> {
             // ensures that the lock is released before the database pointer is dropped.
             transmute::<RwLockReadGuard<'_, ()>, RwLockReadGuard<'static, ()>>(database.schema_txn_lock.read().unwrap())
         };
-        let schema = database.schema.read().unwrap().clone();
+        let schema = database.schema.read().unwrap();
 
         let snapshot: WriteSnapshot<D> = database.storage.clone().open_snapshot_write();
         let type_manager = Arc::new(TypeManager::new(
             database.definition_key_generator.clone(),
             database.type_vertex_generator.clone(),
-            None,
+            Some(schema.type_cache.clone()),
         )); // TODO pass cache
         let thing_manager = ThingManager::new(database.thing_vertex_generator.clone(), type_manager.clone());
 
-        Self {
-            snapshot,
-            type_manager,
-            thing_manager,
-            _schema: schema,
-            _schema_txn_guard: schema_txn_guard,
-            _database: database,
-        }
+        drop(schema);
```

Comment:
How come we have to explicitly drop? Won't it happen automatically on return?

Reply:
It's borrowing `database` making it so it can't be moved into `Self {}` at the end of the scope.